### PR TITLE
Adding kafka layer

### PR DESCRIPTION
**Pull request by ktsakalozos**

### CODE REVIEW

**Comment by pengale** [comment]
I tested this out with my zookeeper charm, and got an error.

Here is what I did:

1) Checked out and built zookeeper from my PR: https://github.com/juju-solutions/bigtop/pull/8
2) Added my quorum patch to the zookeeper charm's resources dir: https://github.com/juju-solutions/layer-apache-bigtop-base/pull/18/files
3) Did the same thing with kafka (pulled the branch, built it, applied the patch).
4) Deployed openjdk and both charms to my local LXD controller, and added a relation between kafka and zookeeper.
5) I got the following error in the kafka charm:

```
2016-06-07 14:32:48 INFO zookeeper-relation-changed Traceback (most recent call last):
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/var/lib/juju/agents/unit-kafka-0/charm/hooks/zookeeper-relation-changed", line 19, in <module>
2016-06-07 14:32:48 INFO zookeeper-relation-changed     main()
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/__init__.py", line 73, in main
2016-06-07 14:32:48 INFO zookeeper-relation-changed     bus.dispatch()
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 421, in dispatch
2016-06-07 14:32:48 INFO zookeeper-relation-changed     _invoke(other_handlers)
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 404, in _invoke
2016-06-07 14:32:48 INFO zookeeper-relation-changed     handler.invoke()
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 280, in invoke
2016-06-07 14:32:48 INFO zookeeper-relation-changed     self._action(*args)
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/var/lib/juju/agents/unit-kafka-0/charm/reactive/kafka.py", line 26, in configure_kafka
2016-06-07 14:32:48 INFO zookeeper-relation-changed     kafka.configure_kafka(zks)
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "lib/charms/layer/bigtop_kafka.py", line 37, in configure_kafka
2016-06-07 14:32:48 INFO zookeeper-relation-changed     bigtop.trigger_puppet()
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "lib/charms/layer/apache_bigtop_base.py", line 235, in trigger_puppet
2016-06-07 14:32:48 INFO zookeeper-relation-changed     r'(# )?export JAVA_HOME.*': 'export JAVA_HOME={}'.format(java_home),
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/usr/local/lib/python3.4/dist-packages/jujubigdata/utils.py", line 195, in re_edit_in_place
2016-06-07 14:32:48 INFO zookeeper-relation-changed     with Path(filename).in_place(encoding=encoding) as (reader, writer):
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/usr/lib/python3.4/contextlib.py", line 59, in __enter__
2016-06-07 14:32:48 INFO zookeeper-relation-changed     return next(self.gen)
2016-06-07 14:32:48 INFO zookeeper-relation-changed   File "/usr/local/lib/python3.4/dist-packages/path.py", line 1450, in in_place
2016-06-07 14:32:48 INFO zookeeper-relation-changed     os.rename(self, backup_fn)
2016-06-07 14:32:48 INFO zookeeper-relation-changed FileNotFoundError: [Errno 2] No such file or directory: Path('/etc/default/bigtop-utils') -> Path('/etc/defaul
t/bigtop-utils.bak')
```

**Comment by pengale** [comment]
I got the same error running bundletester on the charm, built, with patch, as per the above.

**Comment by ktsakalozos** [comment]
Thank you for the review Pete. Managed to reproduce the problem you are reporting. Fixed it with a rebase of this branch. Please, re-review.

**Comment by pengale** [comment]
I have a failing test when running a fresh build of the Kafka charm against a fresh model in my lxd environment on Ubuntu Xenial. Here's the text of the message:

```
ERROR: kafka::01-deploy.py
[/tmp/bundletester-vwtqIt/kafka/tests/01-deploy.py exit 1]
2016-06-10 09:14:14 Starting deployment of local.lxd-test:bilbo
2016-06-10 09:14:15 Deploying services...
2016-06-10 09:14:15  Deploying service jdk using cs:trusty/openjdk-4
2016-06-10 09:14:17  Deploying service kafka using /tmp/charmlh2ia0xl/trusty/kafka
2016-06-10 09:14:20  Deploying service zk using cs:trusty/apache-zookeeper-3
2016-06-10 09:15:28 Config specifies num units for subordinate: jdk
2016-06-10 09:19:06 Adding relations...
2016-06-10 09:19:06  Adding relation kafka:zookeeper <-> zk:zkclient
2016-06-10 09:19:06  Adding relation kafka:java <-> jdk:java
2016-06-10 09:20:16 Deployment complete in 362.47 seconds
/home/petevg/.local/lib/python3.5/site-packages/path.py:1717: DeprecationWarning: path is deprecated. Use Path instead.
  warnings.warn(msg, DeprecationWarning)
F
======================================================================
FAIL: test_deploy (__main__.TestDeploy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/bundletester-vwtqIt/kafka/tests/01-deploy.py", line 30, in test_deploy
    assert 'Kafka' in output, "Kafka daemon is not started"
AssertionError: Kafka daemon is not started

----------------------------------------------------------------------
Ran 1 test in 449.967s

FAILED (failures=1)

PASS: 2 ERROR: 1 Total: 3 (456.214629 sec)
```

**Comment by pengale** [comment]
Tests aside, I was able to do "juju add-relation kafka zookeeper" without errors. :-)

**Comment by ktsakalozos** [comment]
This bug in juju core blocks this charm. Kafka service will not start because its fqdn is not resolvable.
https://bugs.launchpad.net/juju-core/+bug/1593185

File: bigtop-packages/src/charm/kafka/layer-kafka/layer.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+repo: git@github.com:juju-solutions/layer-apache-bigtop-kafka.git
+includes:
+  - 'layer:apache-bigtop-base'
+  - 'interface:zookeeper'
+  - 'interface:kafka'
+options:
+  apache-bigtop-base:
+    packages:
+      - default-jre-headless
```
Comment (kwmonroe) [inline bigtop-packages/src/charm/kafka/layer-kafka/layer.yaml]: we don't need this, right?

Comment (ktsakalozos) [inline bigtop-packages/src/charm/kafka/layer-kafka/layer.yaml]: Yes you are right. We do not need this. Updated the PR and retested.

**Comment by ktsakalozos** [comment]
Closing the PR and submitting upstream.

**Pull request closed without merging**